source/_posts/Docker-Swarm.md

```diff
@@ -0,0 +1,346 @@
+- Het beheren van de containers moet gemakkelijk en overzichtelijk zijn.
+- Aangezien ik een echte Hollander ben moeten de kosten ook een beetje beperkt blijven.
+
+De oplossing welke ik heb gemaakt is een combinatie van Docker Swarm, Traefik, Let's Encrypt en Portainer.
+
+Wat is Docker Swarm
+---
+Docker Swarm is een tool waarmee je Docker-containers kunt beheren en schalen. Als je Docker installeert dan krijg je daar meteen Swarm bij. Docker Swarm is een standaard product van Docker wat bij elke installatie van Docker wordt meegeleverd. Met Docker Swarm kun je een cluster bouwen van verschillende virtuele machines deze worden hierna nodes genoemd. Hierop kunnen de Docker-containers worden gedeployed als een stack\*\* of een service\*. Als je een cluster wilt hebben wat een hoge beschikbaarheid heeft wordt aangeraden om minstens 3 manager nodes te hebben. Docker Swarm maakt namelijk gebruik van het Raft Consensus Algoritme, 1 manager is de leider van de swarm en de status van de manager wordt gesynchroniseerd over de overige managers. Mocht de leider niet meer beschikbaar zijn om wat voor reden dan ook kan een andere manager zijn taken over nemen.
+
+Om te berekenen hoeveel managers er mogen uitvallen voordat het cluster niet meer kan functioneren wordt de volgende berekening gebruikt: (N-1/2).
+Bij dus een cluster van 3 manager mag er 1 manager uitvallen en zal het cluster nog steeds functioneren. Docker adviseert om niet meer dan 7 managers te gebruiken om performance issues met het synchroniseren te voorkomen. Voor meer informatie zie https://docs.docker.com/engine/swarm/raft/
+
+Om een Docker Swarm cluster op te zetten kun je de volgende stappen uitvoeren:
+- Installeer 5 virtuele machine met bijv. Ubuntu server. Voor de 3 managers hebben we niet hele zware virtuele machines nodig. Een basic A1 1.75 GB RAM volstaat al voor een manager node. 
+Voor de 2 worker nodes zou ik kiezen voor een virtuele machine met 8 GB RAM. 
+- Installeer Docker community edition op alle 5 de servers. Volg de handleiding op https://docs.docker.com/install/linux/docker-ce/ubuntu/. 
+- Op 1 van de manager servers voer het volgende commando uit om Docker Swarm te initialiseren.
+```
+$ docker swarm init --advertise-addr "public ipadres van de server"
+Swarm initialized: current node (bvz81updecsj6wjz393c09vti) is now a manager.
+
+To add a worker to this swarm, run the following command:
+
+    docker swarm join \
+    --token SWMTKN-1-3pu6hszjas19xyp7ghgosyx9k8atbfcr8p2is99znpy26u2lkl-1awxwuwd3z9j1z3puu7rcgdbx \
+    172.17.0.2:2377
+
+To add a manager to this swarm, run 'docker swarm join-token manager' and follow the instructions.
+```
+- Voer het bovenstaande docker swarm join token uit op de 2 nodes. 
+- Voer het comamndo docker swarm join-token manager uit en voer het join commando uit op de overige 2 managers.
+
+Als dit klaar gedaan is heb je een Docker Swarm cluster gemaakt zoals in het onderstaande overzicht is weergegeven.
```

Comment:
Kan je niet voor alle machines in cluster een commando uitvoeren? Nu moet je dus op zo'n VM inloggen om een commande uit te voeren?

Reply:
commando* :)

Reply:
Klopt op elke machine moet je inloggen en het join token uitvoeren